pkg/checkpoint/manifest.go

```diff
@@ -73,3 +69,11 @@ func writeManifestIfDifferent(path, name string, data []byte) (bool, error) {
 	glog.Infof("Writing manifest for %q to %q", name, path)
 	return true, writeAndAtomicRename(path, data, 0644)
 }
+
+func writeAndAtomicRename(path string, data []byte, perm os.FileMode) error {
+	tmpfile := filepath.Join(filepath.Dir(path), "."+filepath.Base(path))
```

Comment:
Now I noticed that maybe we should not write the file under the directory.
Even with the prefix of ".", it will still be scanned by `ReadDir()` in `func getFileCheckpoints()`.

Reply:
@yifan-gu since the checkpointer is single-threaded `getFileCheckpoints()` should not run concurrently. The only problem is if the program crashes in the middle of `writeAndAtomicRename` and we have a leftover `.foo.json` file sitting there.

I can adjust `getFileCheckpoints()` to ignore files that start with `.`, WDYT?

Reply:
I guess my question is why we are not just using a system temp dir here?

Reply:
@yifan-gu Atomic renames are only guaranteed in the same directory.

Reply:
@yifan-gu added some code in `getFileCheckpoints()` to clean up old files that start with `.`

Reply:
ok, I didn't found the where said it's only guaranteed in the same directory, but I am fine with the new commit.

Reply:
https://godoc.org/os#Rename calls `rename()` says 'OS-specific restrictions may apply when oldpath and newpath are in different directories. If there is an error, it will be of type *LinkError.'

If you look at http://man7.org/linux/man-pages/man2/rename.2.html, `EXDEV` occurs if the old and new locations are on different filesystems (I've seen this happen before). It's safer to keep it in the same directory.

Reply:
ioutil.TempFile would be a smart choice to create a temporary file within a certain directory

Reply:
@rphillips the only difference would be that it will choose a random rather than deterministic filename, as we would be calling it with the current directory and `.` prefix, and then writing to it... I can make the change anyway though.

Reply:
it isn't a hard blocker... just a suggestion if we need to enumerate those files with a better prefix in the future...

Reply:
Eh already changed it.